src/libmerklecpp/merklecpp.h

```diff
@@ -0,0 +1,1294 @@
+        bytes[i] = tmp;
+      }
+    }
+
+    HashT<SIZE>(const std::vector<uint8_t>& bytes)
+    {
+      if (bytes.size() < SIZE)
+        throw std::runtime_error("not enough bytes");
+      deserialise(bytes);
+    }
+
+    HashT<SIZE>(const std::vector<uint8_t>& bytes, size_t& position)
+    {
+      if (bytes.size() - position < SIZE)
+        throw std::runtime_error("not enough bytes");
+      deserialise(bytes, position);
+    }
+
+    HashT<SIZE>(const std::array<uint8_t, SIZE>& bytes)
+    {
+      std::copy(bytes.data(), bytes.data() + SIZE, this->bytes);
+    }
+
+    size_t size() const
+    {
+      return SIZE;
+    }
+
+    size_t byte_size() const
+    {
+      return SIZE;
+    }
+
+    std::string to_string(size_t num_bytes = SIZE) const
+    {
+      size_t num_chars = 2 * num_bytes;
+      std::string r(num_chars, '_');
+      for (size_t i = 0; i < num_bytes; i++)
+        snprintf(r.data() + 2 * i, num_chars + 1 - 2 * i, "%02X", bytes[i]);
+      return r;
+    }
+
+    HashT<SIZE> operator=(const HashT<SIZE>& other)
+    {
+      std::copy(other.bytes, other.bytes + SIZE, bytes);
+      return *this;
+    }
+
+    bool operator==(const HashT<SIZE>& other) const
+    {
+      return memcmp(bytes, other.bytes, SIZE) == 0;
+    }
+
+    bool operator!=(const HashT<SIZE>& other) const
+    {
+      return memcmp(bytes, other.bytes, SIZE) != 0;
+    }
+
+    void serialise(std::vector<uint8_t>& buffer) const
+    {
+      for (auto& b : bytes)
+        buffer.push_back(b);
+    }
+
+    void deserialise(const std::vector<uint8_t>& buffer, size_t& position)
+    {
+      if (buffer.size() - position < SIZE)
+        throw std::runtime_error("not enough bytes");
+      for (size_t i = 0; i < sizeof(bytes); i++)
+        bytes[i] = buffer[position++];
+    }
+
+    void deserialise(const std::vector<uint8_t>& buffer)
+    {
+      size_t position = 0;
+      deserialise(buffer, position);
+    }
+
+    operator std::vector<uint8_t>() const
+    {
+      std::vector<uint8_t> bytes;
+      serialise(bytes);
+      return bytes;
+    }
+  };
+
+  template <
+    size_t HASH_SIZE,
+    void (*HASH_FUNCTION)(
+      const HashT<HASH_SIZE>& l,
+      const HashT<HASH_SIZE>& r,
+      HashT<HASH_SIZE>& out)>
+  class PathT
+  {
+  public:
+    typedef enum
+    {
+      PATH_LEFT,
+      PATH_RIGHT
+    } Direction;
+
+    typedef struct
+    {
+      HashT<HASH_SIZE> hash; // This is a copy; do we want a shared_ptr<HashT>?
+      Direction direction;
+    } Element;
+
+    PathT(
+      const HashT<HASH_SIZE>& leaf,
+      size_t leaf_index,
+      std::list<Element>&& elements,
+      size_t max_index) :
+      _leaf(leaf),
+      _leaf_index(leaf_index),
+      _max_index(max_index),
+      elements(elements)
+    {}
+    PathT(const PathT& other)
+    {
+      _leaf = other._leaf;
+      elements = other.elements;
+    }
+    PathT(PathT&& other)
+    {
+      _leaf = std::move(other._leaf);
```

Comment:
Does `HashT` need a move constructor?

Reply:
I don't think so. HashT has no members that need to be moved, so it would be the same as the copy constructor. But I can add one if you think it's useful.